src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -2,18 +2,174 @@
+                if (currentNode != null) {
+                    while (currentNode != null) {
+                        if (isEqual(value, currentNode.value)) {
+                            return true;
+                        }
+                        currentNode = currentNode.next;
+                    }
+                }
+            }
+        }
+        return false;
+    }
+
+    private int hash(Object key) {
+        int hash;
+        return (key == null) ? 0 : (hash = key.hashCode()) ^ (hash >>> 16);
+    }
+
+    private int calculateIndex(Object key) {
+        return hash(key) & (table.length - 1);
+    }
+
+    private void resize() {
+        Node<K, V>[] oldTable = table;
+        Node<K, V>[] newTable = new Node[table.length << 1];
+        table = newTable;
+        for (int i = 0; i < oldTable.length; i++) {
+            while (oldTable[i] != null) {
+                Node<K, V> newNode = new Node<>(oldTable[i].key, oldTable[i].value);
+                int index = calculateIndex(newNode.key);
+                if (table[index] == null) {
+                    table[index] = newNode;
+                } else {
+                    Node<K, V> currentNode = table[index];
+                    while (currentNode.next != null) {
+                        currentNode = currentNode.next;
+                    }
+                    currentNode.next = newNode;
+                }
+                oldTable[i] = oldTable[i].next;
+            }
+            oldTable[i] = null;
+        }
+    }
+
+    private boolean isEqual(Object a, Object b) {
+        return (a == b || a != null && a.equals(b));
+    }
+
+    private Node<K, V> removeNode(Object key) {
+        Node<K, V> removedNode;
+        int index = calculateIndex(key);
+        if (table[index] != null) {
+            if (isEqual(key, table[index].key)) {
+                removedNode = table[index];
+                table[index] = table[index].next;
+                size--;
+                return removedNode;
+            } else {
+                Node<K, V> currentNode = table[index];
+                while (currentNode.next != null) {
+                    if (currentNode.next.next != null) {
+                        if (isEqual(key, currentNode.next.key)) {
+                            removedNode = currentNode.next;
+                            currentNode.next = currentNode.next.next;
+                            size--;
+                            return removedNode;
+                        }
+                    } else {
+                        if (isEqual(key, currentNode.next.key)) {
+                            removedNode = currentNode.next;
+                            currentNode.next = null;
+                            size--;
+                            return removedNode;
+                        }
+                    }
+                    currentNode = currentNode.next;
+                }
+            }
+        }
+        return null;
```

Comment:
```suggestion
        int index = calculateIndex(key);
        Node<K, V> node = table[index];

        if (node == null) {
            return null;
        }

        Node<K, V> prev = null;

        while (node != null) {
            if (isEqual(node.key, key)) {
                if (prev == null) {
                    table[index] = node.next;
                } else {
                    prev.next = node.next;
                }
                size--;
                return node;
            }
            prev = node;
            node = node.next;
        }
        return null;
```
can be simplified a little bit

Reply:
how smart) 
I didn't have in mind possibility of additional variable prev. Thanks.

Reply:
ask questions like this in jv_chat